### PR TITLE
Fix handling of environment variables with empty string values.

### PR DESCRIPTION
**Pull request by adityamandaleeka**
Fix https://github.com/dotnet/corefx/issues/17614

[`GetEnvironmentVariableCore`](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Environment.cs#L679) doesn't handle the case where an environment variable is found but has length 0. If this happens, it effectively just does `StringBuilderCache.Acquire` followed by `StringBuilderCache.GetStringAndRelease`.

@JeremyKuhne @jkotas

### CODE REVIEW

**Comment by adityamandaleeka** [comment]
@stephentoub `GetEnvironmentVariableW` in the PAL doesn't touch the buffer passed in unless the variable is found and is non-empty.

The other issue here is that `GetEnvironmentVariableW` uses a return value of 0 to indicate that it failed as well. I should also verify that the last error is ERROR_SUCCESS before returning empty string; I'll update the PR.

**Comment by omajid** [comment]
What does the windows implementation of `GetEnvironmentVariable` do? Does it touch the buffer if the variable is found and empty? In other words, should the pal implementation of `GetEnvironmentVariable` be modified?

**Comment by adityamandaleeka** [comment]
@omajid I don't think you can have empty variables in Windows. `foo=` will just unset foo.

**Comment by stephentoub** [comment]
> GetEnvironmentVariableW in the PAL doesn't touch the buffer passed in unless the variable is found and is non-empty

And that's how the Win32 function behaves on Windows? I'd have guessed that's the bug, in the PAL. Otherwise, why does it work on Windows but fail on Unix?

**Comment by omajid** [comment]
Thanks!

**Comment by adityamandaleeka** [comment]
@stephentoub Correct me if I'm wrong, but as I said above, I don't think you can have empty environment variables on Windows. On Unix, `unset foo` (or just never setting `foo`) and `foo=` are distinct.

The former will leave no entry for `foo` in the environment block, while the latter will have `foo=` in the block. Also, `getenv` will return a null pointer for the former and a pointer to an empty string for the latter. This is the behavior we're trying to match in the PAL.

**Comment by stephentoub** [comment]
> I don't think you can have empty environment variables on Windows

What happens if you call the Win32 SetEnvironmentVariable with an empty string value (rather than NULL, which would delete it)?

**Comment by adityamandaleeka** [comment]
@stephentoub Thanks, that works. It looks like the `GetEnvironmentVariable` return value is the same whether a variable is not found or is found but empty (it returns 0 in both cases). However, in the empty case, it returns `ERROR_SUCCESS` and sets the value in the buffer and if it's not found, it returns `ERROR_ENVVAR_NOT_FOUND` and the buffer is unmodified.

I'll send another PR with the change in the PAL to make GetEnvironmentVariable do the same thing.

**Comment by dscho** [comment]
Sorry to revive an old, closed thread, but...

> It looks like the `GetEnvironmentVariable` return value is the same whether a variable is not found or is found but empty (it returns 0 in both cases). However, in the empty case, it returns `ERROR_SUCCESS` and sets the value in the buffer and if it's not found, it returns `ERROR_ENVVAR_NOT_FOUND` and the buffer is unmodified.

I just found out the hard way that there is a bug: if you call `GetEnvironmetVariable()` for an unset environment variable, it 0 and `GetLastError()` returns `ERROR_ENVVAR_NOT_FOUND` as expected, but if you then *immediately* call the function again, this time for an environment variable with an empty value, then it returns 0 but `GetLastError()` *still returns `ERROR_ENVVAR_NOT_FOUND`*!

In other words, it seems that you have to call `SetLastError(ERROR_SUCCESS);` before calling `GetEnvironmentVariableW()` if you care about distinguishing between empty values and unset variables.

Maybe this needs to be fixed in PAL, too?

I am thinking that this `SetLastError()` call would need to go here: https://github.com/dotnet/coreclr/blob/91e1ffccc38fc87f6f496eb056396a1b775f08af/src/System.Private.CoreLib/shared/Interop/Windows/Kernel32/Interop.GetEnvironmentVariable.cs#L16 (but I am not familiar with the .NET Core source code...)

**Comment by jkotas** [comment]
@dscho If you believe that there is a real bug in .NET Core, please open a new issue with a code to reproduce the problem.

**Comment by dscho** [comment]
@jkotas I only have code in C that demonstrates the problem (I only stumbled across this here ticket when I was looking for more information, and figured I'd let you gentle people know about this issue).
